src/client/dfs/dcache.c

```diff
@@ -904,48 +993,318 @@ drec_del_at_dact(dfs_dcache_t *dcache, dfs_obj_t *rec)
+		return daos_der2errno(rc);
+
+	D_ALLOC_PTR(dcache_tmp);
+	if (dcache_tmp == NULL)
+		D_GOTO(err_shm, rc = ENOMEM);
+
+	dcache_tmp->dd_dfs  = dfs;
+	dcache_tmp->dd_type = DFS_CACHE_SHM;
+
+	/** calculate a hash with the pool and cont uuid */
+	rc = dc_pool_hdl2uuid(dfs->poh, NULL, &pool_cont_uuid[0]);
+	if (rc)
+		D_GOTO(err_dcache, rc = daos_der2errno(rc));
+
+	rc = dc_cont_hdl2uuid(dfs->coh, NULL, &pool_cont_uuid[1]);
+	if (rc)
+		D_GOTO(err_dcache, rc = daos_der2errno(rc));
+
+	/* acquire the pointer of the existing shm LRU cache for dentry */
+	dcache_tmp->shm.d_shm_lru_dentry = shm_lru_get_cache(CACHE_DENTRY);
+	D_ASSERT(dcache_tmp->shm.d_shm_lru_dentry != NULL);
+	dcache_tmp->shm.dd_pool_cont_hash =
+	    d_hash_murmur64((const unsigned char *)pool_cont_uuid, sizeof(uuid_t) * 2, 0);
+
+	rc = dcache_add_root(dcache_tmp);
+	if (rc != 0)
+		D_GOTO(err_dcache, rc = daos_der2errno(rc));
+
+	dfs->dcache = dcache_tmp;
+	return 0;
+
+err_dcache:
+	D_FREE(dcache_tmp);
+err_shm:
+	shm_fini();
+	return rc;
+}
+
 int
-dcache_create(dfs_t *dfs, uint32_t bits, uint32_t rec_timeout, uint32_t gc_period,
+dcache_create(dfs_t *dfs, int type, uint32_t bits, uint32_t rec_timeout, uint32_t gc_period,
 	      uint32_t gc_reclaim_max)
 {
 	D_ASSERT(dfs);
 	if (bits == 0)
 		return dcache_create_dact(dfs);
+	if (type == DFS_CACHE_DRAM)
+		return dcache_create_act(dfs, bits, rec_timeout, gc_period, gc_reclaim_max);
 
-	return dcache_create_act(dfs, bits, rec_timeout, gc_period, gc_reclaim_max);
+	return dcache_create_shm(dfs);
 }
 
 int
 dcache_destroy(dfs_t *dfs)
 {
 	D_ASSERT(dfs->dcache != NULL);
-	D_ASSERT(dfs->dcache->destroy_fn != NULL);
+	if (dfs->dcache->dd_type == DFS_CACHE_SHM) {
+		D_FREE(dfs->dcache);
+		return 0;
+	}
+	D_ASSERT(dfs->dcache->dh.destroy_fn != NULL);
 
-	return dfs->dcache->destroy_fn(dfs->dcache);
+	return dfs->dcache->dh.destroy_fn(dfs->dcache);
 }
 
 int
-dcache_find_insert(dfs_t *dfs, char *path, size_t path_len, int flags, dfs_obj_t **rec,
+dcache_find_insert(dfs_t *dfs, char *path, size_t path_len, int flags, dfs_obj_t **_rec,
 		   mode_t *mode, struct stat *stbuf)
 {
-	D_ASSERT(rec != NULL);
+	dfs_obj_t      *rec;
+	dfs_obj_t      *parent;
+	char            key[SHM_DCACHE_KEY_PREF_SIZE + DFS_MAX_NAME];
+	char           *name;
+	size_t          name_len;
+	bool            skip_stat = false;
+	int             rc;
+	uint64_t       *ptr_key = (uint64_t *)key;
+	size_t          key_len;
+	shm_lru_node_t *node_found = NULL;
+	char           *value;
+	size_t          val_size;
+
+	D_ASSERT(_rec != NULL);
 	D_ASSERT(dfs->dcache != NULL);
 	D_ASSERT(path != NULL);
-	D_ASSERT(dfs->dcache->find_insert_fn != NULL);
 
-	return dfs->dcache->find_insert_fn(dfs->dcache, path, path_len, flags, rec, mode, stbuf);
+	if (dfs->dcache->dd_type == DFS_CACHE_DRAM) {
+		D_ASSERT(dfs->dcache->dh.find_insert_fn != NULL);
+		return dfs->dcache->dh.find_insert_fn(dfs->dcache, path, path_len, flags, _rec,
+						      mode, stbuf);
+	}
+
+	name       = path;
```

Comment:
Shoud it not be possible to use the `find_insert_fn` callback for the `DFS_CACHE_SHM` as it is done for the `DFS_CACHE_DRAM` ?
From my point of view it will make the code more clear.